specs/anchored-region/anchored-region.md

```diff
@@ -0,0 +1,260 @@
+    <anchored-region
+        anchor="anchor"
+        verticalpositioningmode="onetime"
+    >
+      This shows up above or below the anchor depending on available space
+    </anchored-region>
+    ...stuff...
+</div>
+
+```
+
+A region that overlaps the anchor and renders above or below it depending on available space.
+
+```
+<div id="viewport">
+    ...stuff...
+    <button id="anchor">
+        Button is an anchor
+    </button>
+    <anchored-region
+        anchor="anchor"
+        verticalPositioningMode="dynamic"
+        verticalinset="true"
+    >
+      This overlaps the anchor and extends above or below it depending on available space.
+    </anchored-region>
+    ...stuff...
+</div>
+
+```
+
+A region renders above or below the anchor based on available space and is sized to match the available.
+
+```
+<div id="viewport">
+    ...stuff...
+    <button id="anchor">
+        Button is an anchor
+    </button>
+    <anchored-region
+        anchor="anchor"
+        verticalPositioningMode="dynamic"
+        verticalscalingenabled="true"
+    >
+       This region renders above or below the anchor based on available space and sizes itself to match.
+    </anchored-region>
+    ...stuff...
+</div>
+
+```
+
+A region that renders below the anchor until that space is less than 100px.
+```
+<div id="viewport">
+    ...stuff...
+    <button id="anchor">
+        Button is an anchor
+    </button>
+    <anchored-region
+        anchor="anchor"
+        verticalpositioningmode="dynamic"
+        verticaldefaultposition="below"
+        verticalthreshold="100"
+    >
+      This shows renders below the anchor as long as there is at least 100px available there.
+    </anchored-region>
+    ...stuff...
+</div>
+```
+
+The dimensions of the anchored region will match the dimensions of the content unless scaling is enabled on a particular axis (verticalscalingenabled & horizontalscalingenabled) in which case it will fill all available space between the anchor and viewport.
+
+The component allows users to set a "Positioning Mode" on each axis which defines how the component will behave:
+- 'uncontrolled':  The anchored region will appear as it normally would in document flow.
+- 'locktodefault': The anchored region will always be placed in the specified default position regardless of available space.  For example a menu that always opens upwards.
+- 'dynamic': The anchored region is placed relative to the anchor element based on how much space is available between it and the viewport.  For example a menu that opens opens upwards when it is near the bottom of the page, and downwards when near the top.
+- 'onetime': As the 'dynamic' option, but position is not changed after initial placement.
+
+### API
+
+*Component name:*
+- `fast-anchored-region`
+
+*Attributes:*
+- anchor - The html id of the HTMLElement used as the anchor around which the positioning region is placed.  This must be set for the component's positioning logic to be active.
```

Comment:
Should this be an ID or should it be a slot? I could forsee an implementation that was more encapsulated looking like this:
```
<fast-anchored-region>
    <button slot="anchor">My Anchor</button>
    <div>The content of the anchored region</div>
</fast-anchored-region>
```

Reply:
No, the anchor by definition can't be a child of the anchored region.  The pattern you show would likely be what you would see for a select like component that wraps the anchor (ie the trigger button that opens the select) as well as an anchored-region (the menu).

Reply:
But the region we're talking about could be the slotted region, there doesn't need to be a wrapper if you don't want to have one. Correct me if I'm wrong but couldn't you do something like this:

```
<slot name="anchor">
    <button>My default anchor</button>
</slot>
<div part="content">
    <slot>
        <div>My default content</div>
    </slot>
</div>
```

Others might have a better idea of how to do this but you can see a similar concept in the checkbox spec https://github.com/microsoft/fast-dna/blob/master/specs/checkbox.md

Reply:
The pattern you describe is pretty close to what I'd expect for a Select component that used this anchor to place its menu.  However I don't think it makes sense to implement that at the level of this component because there are common scenarios, like a flyout menu for example, where a single menu instance exists at the page level and the developer has it appear next to different anchor elements as the user interacts.  In the proposed API this would be done by changing the anchor id, but doesn't really work with the slot scenario.

Reply:
@scomea I think there could be templating solutions though to handle what element is projected or assigned to the slot. If this region is generating DOM and the element needs to be on the same level as the region - it seems like the safest way to ensure the markup is as expected would be to leverage slots. I may also be misunderstanding something though.

Reply:
I think using slots for this too prescriptive.  What if I want to attach a flyout to an element that I don't have control over rendering? Why should this element care *where* the anchor is? What is the *benefit* or using a slot vs an id reference?

I don't think an anchored region should care what it is anchored to and how that anchor is rendered - all it really should care about is if that element exists.